### PR TITLE
net/devif_poll: optimize device buffer alloc in txpoll

### PR DESCRIPTION
**Pull request by anchao**
## Summary

1. net/devif_poll: optimize device buffer alloc in txpoll

Allocate the device buffer only if the protocol really need to send data.
not all protocols require the driver to prepare additional iob before
sending, especially UDP, each iob reserves l2/l3 header in advance
after prepare write buffer, net device could reuse this entry to send directly

Signed-off-by: chao an <anchao@xiaomi.com>

2. net/devif: reuse devif_send() for can/pkt/icmp stack

## Impact

N/A

## Testing

iperf udp test

### CODE REVIEW

File: net/udp/udp_sendto_unbuffered.c [hunk 1]
```diff
@@ -192,6 +192,15 @@ static uint16_t sendto_eventhandler(FAR struct net_driver_s *dev,
 
       else
         {
+          /* Copy the user data into d_appdata and send it */
+
+          devif_send(dev, pstate->st_buffer,
```
Comment (xiaoxiang781216) [inline net/udp/udp_sendto_unbuffered.c:197]: why change the order

File: net/udp/udp_sendto_unbuffered.c [hunk 2]
```diff
@@ -202,15 +211,6 @@ static uint16_t sendto_eventhandler(FAR struct net_driver_s *dev,
           sendto_ipselect(dev, pstate);
 #endif
 
-          /* Copy the user data into d_appdata and send it */
-
-          devif_send(dev, pstate->st_buffer,
-                     pstate->st_buflen, udpip_hdrsize(pstate->st_conn));
-          if (dev->d_sndlen == 0)
-            {
-              return flags;
```
Comment (xiaoxiang781216) [inline net/udp/udp_sendto_unbuffered.c:211]: remove the extra space at line 219

File: net/udp/udp_devpoll.c [hunk 3]
```diff
@@ -93,17 +93,6 @@ void udp_poll(FAR struct net_driver_s *dev, FAR struct udp_conn_s *conn)
...
-       * may not actually be used.
-       */
-
-#if defined(CONFIG_NET_IPv4)
-      udp_ipv4_select(dev);
-#else /* if defined(CONFIG_NET_IPv6) */
-      udp_ipv6_select(dev);
-#endif
-
       dev->d_len     = 0;
```
Comment (xiaoxiang781216) [inline net/udp/udp_devpoll.c:96]: remove the extra space

File: net/pkt/pkt_poll.c [hunk 4]
```diff
@@ -65,7 +65,6 @@ void pkt_poll(FAR struct net_driver_s *dev, FAR struct pkt_conn_s *conn)
     {
       /* Setup for the application callback */
 
-      dev->d_appdata = IPBUF(0);
       dev->d_len     = 0;
```
Comment (xiaoxiang781216) [inline net/pkt/pkt_poll.c:68]: remove the extra space

File: net/mld/mld_poll.c [hunk 5]
```diff
@@ -60,7 +60,6 @@ void mld_poll(FAR struct net_driver_s *dev)
 
   /* Setup the poll operation */
 
-  dev->d_appdata = IPBUF(IPv6_HDRLEN);
```
Comment (xiaoxiang781216) [inline net/mld/mld_poll.c:63]: remove the extra space

File: net/igmp/igmp_poll.c [hunk 6]
```diff
@@ -151,13 +151,9 @@ static inline void igmp_sched_send(FAR struct net_driver_s *dev,
...
 {
   FAR struct igmp_group_s *group;
-  uint16_t iphdrlen;
 
   /* Setup the poll operation */
 
-  iphdrlen       = IPv4_HDRLEN + RASIZE;
-
-  dev->d_appdata = IPBUF(iphdrlen + IGMP_HDRLEN);
   dev->d_len     = 0;
```
Comment (xiaoxiang781216) [inline net/igmp/igmp_poll.c:157]: remove the extra space

File: net/udp/udp_sendto_unbuffered.c (hunk 1 above)
Comment (anchao) [reply net/udp/udp_sendto_unbuffered.c:197]: New send flow will prepare the d_iob buffer in devif_send(), sendto_ipselect() depends on d_iob to update d_appdata

File: net/icmp/icmp_sendmsg.c [hunk 7]
```diff
@@ -107,25 +115,16 @@ static void sendto_request(FAR struct net_driver_s *dev,
...
-
   /* Initialize the IP header. */
 
   ipv4_build_header(IPv4BUF, dev->d_len, IP_PROTO_ICMP,
                     &dev->d_ipaddr, &pstate->snd_toaddr,
                     IP_TTL_DEFAULT, NULL);
 
+  /* Copy the ICMP header and payload into place after the IPv4 header */
+
+  icmp              = IPBUF(IPv4_HDRLEN);
```
[the remark this answers is not in the document]
Comment (anchao) [reply net/icmp/icmp_sendmsg.c:126]: Done

File: net/icmp/icmp_poll.c [hunk 8]
```diff
@@ -66,7 +66,6 @@ void icmp_poll(FAR struct net_driver_s *dev, FAR struct icmp_conn_s *conn)
 
   /* Setup for the application callback */
 
-  dev->d_appdata = IPBUF(IPICMP_HDRLEN);
   dev->d_len     = 0;
```
[the remark this answers is not in the document]
Comment (anchao) [reply net/icmp/icmp_poll.c:69]: Done

File: net/pkt/pkt_poll.c (hunk 4 above)
Comment (anchao) [reply net/pkt/pkt_poll.c]: Done

File: net/mld/mld_poll.c (hunk 5 above)
Comment (anchao) [reply net/mld/mld_poll.c:63]: Done

File: net/udp/udp_devpoll.c (hunk 3 above)
Comment (anchao) [reply net/udp/udp_devpoll.c]: Done

File: net/udp/udp_sendto_unbuffered.c (hunk 2 above)
Comment (anchao) [reply net/udp/udp_sendto_unbuffered.c:211]: Done

File: net/icmp/icmp_sendmsg.c [hunk 9]
```diff
@@ -107,25 +115,16 @@ static void sendto_request(FAR struct net_driver_s *dev,
...
+
   /* Calculate the ICMP checksum. */
 
   icmp->icmpchksum  = 0;
```
Comment (xiaoxiang781216) [inline net/icmp/icmp_sendmsg.c:130]: remove the extra spae

Comment (anchao) [reply net/icmp/icmp_sendmsg.c]: Done

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**